### ISSUE TITLE
Templating engine compilation caching

### ISSUE THREAD

**Issue by pofider** [jsreport/jsreport#117]
Script manager compiles handlebars/jsrender templates every time. This could be cached.

**Comment by bjrmatos** [comment]
yesterday i was trying to resolve this problem but i was hit by limitations of child process interprocess communication..

What i am trying to do:

1.- add support in `toner` for template caching, basically mantaining a hash of `templateId: engineFn`, on each request will be a check for trying reuse the `engineFn` in the hash, if not present just call the proper engine again. where `templateId` is template's `shortid` or `name`, for now `shortid` takes precedence over `name` because `name` is not a unique value (i can register two or more templates with the same name). 

for future plans #118  (plain files) we need to make sure that template's `name` is a unique value (basically don't allow templates with the same name on save) in order to match the file system behaviour.

another problem i've found is that requests for rendering templates's `header` and `footer` are made with the same `shortid/name` of template.. we need to find a way to differentiate template's rendering request and header/footer template's rendering request in order to do the correct caching (maybe i have missed something, i didn't check if there is a mechanism that differentiate those requests already, i am concentrating in make caching works first).

2.- change engine function contract, for example i will take `toner-handlebars` for reference, the current implementation looks like:

 ```js
module.exports = function (html, helpers, data) {
    var handlebars = require("handlebars");

    if (helpers) {
        for (var h in helpers) {
            if (helpers.hasOwnProperty(h))
                handlebars.registerHelper(h, helpers[h]);
        }
    }

    // replace {#image {{a}}} with {#image {{a}}%}% and back so it compiles
    html = html.replace(/({#[^{}]{0,20} {{{[^{}]{0,20})(}}}})/g, "$1}}}%}%");
    html = html.replace(/({#[^{}]{0,20} {{[^{}]{0,20})(}}})/g, "$1}}%}%");

    var compiledTemplate = handlebars.compile(html);
    return compiledTemplate(data).replace("%}%", "}");
};
```

the new implementation will look like:

```js
module.exports = function (html) {
    var handlebars = require("handlebars");

    // replace {#image {{a}}} with {#image {{a}}%}% and back so it compiles
    html = html.replace(/({#[^{}]{0,20} {{{[^{}]{0,20})(}}}})/g, "$1}}}%}%");
    html = html.replace(/({#[^{}]{0,20} {{[^{}]{0,20})(}}})/g, "$1}}%}%");

    var compiledTemplate = handlebars.compile(html);

    return function(reqHelpers, reqData) {
        // since jsreport only support local helpers (for now), i'm making helpers really local to the template (not using `handlebars.registerHelper`)
        return compiledTemplate(reqData, { helpers: reqHelpers }); // handlebars support local helpers in this way
    };
};
```

we need that engines export a function rather than the final html, this function will be cached in toner's hash as `engineFn`.. **is a breaking change** but we need this to make caching simple and force engine's author to compile templates, another gain is that during jsreport initialization (in production mode) we can make a first call to all engines with all templates and save the (`engineFn`) results in cache, so the first rendering won't be impacted by template's compilation

3.- **(this is the part that have limitations)** pass `engineFn` to `engineScript.js` in toner

so in the ideal world, to make caching works the only thing left is pass template's `engineFn` to `engineScript.js` as a new parameter and call this function inside (`engine(helpers, data)`) and return the result (final html).. unfortunately this don't work because child process limitations, i can't pass a function to `toner.scriptManager.execute`, the function automatically will be converted to `undefined` inside the child process that gets created by [script-manager](https://github.com/pofider/node-script-manager)

i have tried other ways to make caching works but all have the same problem with child process limitations.

in the end i found that using the `vm` module will be secure enough for jsreport use case, we only have to reach a certain level of security for our use case.. only require allowed modules, don't allow main process global/local scope modification to template's helpers, all can be achieved with the `vm` module. (basically what i'm asking you is to not use [script-manager](https://github.com/pofider/node-script-manager) :(, or create a strategy that don't rely on child processes.. without workers only timeout execution)

i know that using another process for template execution is the final way to make it really really secure but i think using `vm` module is enough for us.

do you have another approach?

**Comment by bjrmatos** [comment]
i forgot to mention that cache invalidation will be implemented as `tonerInstance.invalidateCache(templateId)`, this method will be called by [jsreport templates extension](https://github.com/jsreport/jsreport/tree/master/extension/templates) when it detects that some template content has changed.

**Comment by bjrmatos** [comment]
working proof of concept code without child processes:

```js

var vm = require('vm');

var helpers = 'function saysomething(text) { return text; }';

// compile code only one time
var execScript = new vm.Script("respond(null, render(m.template.helpers, m.data))");

function handlebarsEngine(html) {
  var handlebars = require("handlebars");

  var compiledTemplate = handlebars.compile(html);

  return function(reqHelpers, reqData) {
      return compiledTemplate(reqData, { helpers: reqHelpers });
  };
}

var engineFn = handlebarsEngine('<h1>{{text}} - {{saysomething "my name is Boris"}}<h1>');

function executeInSandbox(inputs, done) {
  var _require = function (moduleName) {
    return require(moduleName);
  };

  var sandbox = {
    m: inputs,
    require: _require,
    render: inputs.engine,
    respond: function(err, content) {
      if (err) {
        return done(err);
      }

      done(null, content);
    }
  };

  var sandboxForHelpers = {
    require: _require
  };

  if (inputs.template.helpers) {
      //first grab helpers as it would be an object { "foo" : function... } for back compatibility reasons
      //when its not an object eval again and let helpers register into globals

      vm.runInNewContext("jsrHelpers = " + inputs.template.helpers, sandboxForHelpers);

      if (sandboxForHelpers.jsrHelpers && typeof sandboxForHelpers.jsrHelpers === 'object') {
          inputs.template.helpers = sandboxForHelpers.jsrHelpers;
      } else {
          delete sandboxForHelpers.jsrHelpers;
          vm.runInNewContext(inputs.template.helpers, sandboxForHelpers);

          inputs.template.helpers = {};

          for (var fn in sandboxForHelpers) {
              if (sandboxForHelpers.hasOwnProperty(fn) && typeof sandboxForHelpers[fn] === "function") {
                  inputs.template.helpers[fn] = sandboxForHelpers[fn];
              }
          }
      }
  } else {
    inputs.template.helpers = {};
  }

  try {
    execScript.runInNewContext(sandbox);
  }
  catch(e) {
    if (!e.message) {
        e = new Error(e);
    }

    return console.error(e);
  }
}

executeInSandbox({
  engine: engineFn,
  data: {
    text: 'Hola'
  },
  template: {
    helpers: helpers
  }
}, function(err, content) {
  if (err) {
    return console.error(err);
  }

  console.log('content:', content);
});
```

**Comment by pofider** [comment]
1. Why should we use template id as the cache key? Why not to use triple `html:helpers:engine` instead? Then there is no need for cache invalidation or some extra dealing with headers, footers... The cache invalidation would be anyway difficult to handle in the clustered web farm environment.

2. Yes, returning a function would be required. Doing a breaking change is fine.

3. Yes, I also don't see a way how to pass the compiled function to the child process. However, do we need it... ?

Using `http-server` strategy which reuses worker node instances has likely anyway better performance gains than the whole template caching. And when people care about the performance they don't use `dedicated-process`. So what about wire up the template caching just for the `http-server` strategy? Or more precisely wire it up to both but without actual gains when the node process is created for each rendering.

The difference is that we can hold the cache directly in the worker node processes, because they won't go away, the manager is reusing them. In the `toner/engineScript.js` it could be something like: 

```js
var cache = {};
...
var cacheKey = input.template.html + ':' + input.template.helpers + ':' + input.template.engine;

if (!cache[cacheKey]) {
  //give us the compiled fn
  cache[cacheKey] = require(inputs.engine)(input.template.html, input.template.helpers);
}

//use cache[cacheKey] in the vm to render html from input data
```

**Comment by pofider** [comment]
Just a comment for a rendering strategy without child processes. This is what I have also in the future plans. It would be the most performing strategy for productions where the templates are not modified in the real time. However it is not the highest priority for me because the pdf printing takes 95% of the rendering anyway. And also, I have experienced major issues on production system using jsreport ancestor because of template's endless loops. That's why I even decided to don't implement it until there will be demand for it and rather use worker child processes.

**Comment by bjrmatos** [comment]
> Why should we use template id as the cache key? Why not to use triple html:helpers:engine instead?

`'<h1>...lot of content...</h1>' + ':' + 'function helpers() { ... }' + ':' + 'phantom-pdf'` that seems a little crazy don't you think? 

> Then there is no need for cache invalidation or some extra dealing with headers, footers... The cache invalidation would be anyway difficult to handle in the clustered web farm environment.

i see your point here, but engines are agnostic of recipes so we can re-use cached templates no matter what recipe is being using for rendering, i know cache invalidation would be pain to implement (it is one of the hardest problem in computer science anyway) but i can see ways to handle it.. to not make it very hard at least we can implement a button in studio "compile" or "save for production" or something (enabled only in a live system in production mode), so by default saving the template won't do anything on a live system, we only invalidate the template cache if the user press that button so we don't have to implement complex mechanism to handle cache invalidation..


> Yes, I also don't see a way how to pass the compiled function to the child process. However, do we need it... ?
> 
> Using http-server strategy which reuses worker node instances has likely anyway better performance gains than the whole template caching. And when people care about the performance they don't use dedicated-process. So what about wire up the template caching just for the http-server strategy? Or more precisely wire it up to both but without actual gains when the node process is created for each rendering.

yes, we need it because i think the cache managment (and cache storing) should be the responsability of the toner instance not a responsibility of the rendering execution part, so we can easily support this without actually do any rendering -> `"another gain is that during jsreport initialization (in production mode) we can make a first call to all engines with all templates and save the (engineFn) results in cache, so the first rendering won't be impacted by template's compilation"`

if the dedicated-process strategy is not the most performant, should we make suffer our users more with template compilation on every request? template caching should be a performance gain for every user.

**Comment by bjrmatos** [comment]
> Just a comment for a rendering strategy without child processes. This is what I have also in the future plans. It would be the most performing strategy for productions where the templates are not modified in the real time. However it is not the highest priority for me because the pdf printing takes 95% of the rendering anyway. And also, I have experienced major issues on production system using jsreport ancestor because of template's endless loops. That's why I even decided to don't implement it until there will be demand for it and rather use worker child processes.

fair enough, i think we should implement this now, after all we are in node.js.. where the performance gains should be priority

**Comment by bjrmatos** [comment]
as you previously mentioned implementing cache invalidation in a cluster/multi apps environment will be a pain, in the end i found that probably we didn't even need cache invalidation at all.. i will post (in [jsreport-fs-store](https://github.com/jsreport/jsreport-fs-store)) some thoughts i have about how to deal with this without such complexity.

summary of my proposal for caching:

1.- add support in toner module for template caching, basically mantaining a hash of `templateId: engineFn` (as described in my initial comment) where `templateId` is the name of template.

2.- make template's name a unique value, don't allow templates with the same name (this will be covered without effort when jsreport is based on [jsreport-fs-store](https://github.com/jsreport/jsreport-fs-store), because template's name will be equal to the name of template's directory which is unique in the file system

3.- differentiate rendering requests for template's header/footer, just to cache it separetly (after all it is a rendering request on its own), the implementation would only require to pass an additional boolean value when is header or footer (`header: true` or  `footer: true`)

4.- change engine function contract (as described in my initial comment)

5.- use a template rendering strategy without child processes, probably only using the `vm` module (benefits described in my previous comment)
 
as i mentioned cache invalidation wouldn't be needed (i will post on [jsreport-fs-store](https://github.com/jsreport/jsreport-fs-store) why i think cache invalidation won't be necessary)

these changes should wait until jsreport is completely based on [jsreport-fs-store](https://github.com/jsreport/jsreport-fs-store)

**Comment by pofider** [comment]
I am still not convinced about using `templateId=name` as the cache key. I proposed the key to be template html, helpers and engine. This is also recipe agnostic. Didn't you misread it ? Getting the compiled function from the cache is just about comparing strings. Are you afraid about performance here?  

The cons of using template id as the cache key

1. Doesn't work with anonymous templates
It is very common to use jsreport as stateless rendering service. This means there are no templates stored in jsreport and the external systems are sending anonymous templates inside each rendering request. The cache stops to work here, because there is no template identification. It would work by design if the cache key is based on html, helpers and engine.

2. Cache invalidation and transparency
This would need to be implemented. It is not difficult, we have the template last modification date. However it is just increasing the complexity. There is no need for invalidation when using cache key proposed by me.

3. It doesn't work with child processes. 
For example jsreportonline.net service relies on child processes, because it happens regularly someone has endless loop in the helper. If we use cache key proposed by me, it will work for `http-server` strategy and also for the new `in-process` strategy which won't use child processes.

Please give more comments on this topic...

My proposed solution is to use cache key - html,helpers and engine. 
Also implement the new `in process` strategy without child processes which will be the fastest one. Both `in-process` and `http-server` strategies will benefit from the compiled templates caches.

**Comment by bjrmatos** [comment]
> I am still not convinced about using templateId=name as the cache key. I proposed the key to be template html, helpers and engine. This is also recipe agnostic. Didn't you misread it ?

you're right, i have readed it wrong..

> Getting the compiled function from the cache is just about comparing strings. Are you afraid about performance here?

i'm worried about memory consumption, `html:helpers:engine` is a lot of content to store in an object key and can consume too much memory in a large app with many reports, what about using a simple md5 hash? we can use [hasha](https://github.com/sindresorhus/hasha) (just for a nice crypto API) to calculate a hash of `html:helpers:engine` and use that as the `templateId` (it will look somethink like `655b7e29252ab7924a68b801cbbb9e51`)

for the rest i agree with you, this is the most painless way to handle cache invalidation

**Comment by bjrmatos** [comment]
i was missing the use case for anonymous templates and jsreport as a service, i need to dive more into all parts of jsreport to see the whole picture correctly :)

**Comment by pofider** [comment]
Yes, using hash can be one step further. The final solution should also provide a LRU cache invalidation algorithm.  

I think we are clear here. I submitted also an extra request for the `in process` strategy. #144

**Comment by bjrmatos** [comment]
cool!

> LRU cache invalidation algorithm.

do you have a link with more information about this strategy?

**Comment by bjrmatos** [comment]
never mind i think i have a good link http://mcicpc.cs.atu.edu/archives/2012/mcpc2012/lru/lru.html

**Comment by pofider** [comment]
Ok, I just ment something like this. But I don't insist on particular implementation or solution.

Configuration value for max cache size, default can be lets say 100. To every cache entry we keep the last datetime it was used. If the cache size grows to 101, just remove the last used one.

**Comment by bjrmatos** [comment]
all implemented in this lovely [module](https://github.com/isaacs/node-lru-cache)!

**Comment by pofider** [comment]
Great! It looks very simple, something I expected.

**Comment by pofider** [comment]
This has already shipped to jsreport-core master. It was quite easy to implement thanks to the great @bjrmatos analysis. 

The only difference is that the cache key is built only from `templateContent:engine` because it turned out all the engines actually compile the template only from html and helpers are passed only to already compiled function. In other words, helpers doesn't need to participate in the cache key, changing helper doesn't affect compiled template.

Additionally I was able to make this change back compatible with the current engines implementation. So we can have this in master without breaking the old stuff and change engines step by step. This will come with release 1.

**Comment by pofider** [comment]
This is now released in 1.0.
